legate/core/operation.py

```diff
@@ -399,13 +397,39 @@ def add_scalar_arg(
         ----------
         value : Any
             Scalar value or a tuple of scalars (but no nested tuples)
-        dtype : DType
+        dtype : Dtype
             Data type descriptor for the scalar value. A descriptor ``(T,)``
-            means that the value is a tuple of elements of type ``T``.
+            means that the value is a tuple of elements of type ``T`` (i.e.,
+            equivalent to ``array_type(T, len(value))``).
         """
-        if not isinstance(dtype, (ty.Dtype, tuple)):
-            raise ValueError(f"Unsupported type: {dtype}")
-        self._scalar_args.append((value, dtype))
+        sanitized: ty.Dtype
+        if isinstance(dtype, ty.Dtype):
+            sanitized = dtype
+            if isinstance(sanitized, ty.FixedArrayDtype) and not isinstance(
+                value, (tuple, Shape)
+            ):
+                raise ValueError(f"{value} is not a valid scalar")
+        elif isinstance(dtype, tuple):
+            assert len(dtype) == 1 and isinstance(dtype[0], ty.Dtype)
```

Comment:
Is `add_scalar_arg` a function that users (i.e. devs building tools on `legate.core` in other projects) will use directly? If so, I'd consider `dtype` as a user-supplied value, and therefore not appropriate for an assertion since it is strictly outside out internal control (could raise a `ValueError` instead)

Reply:
yeah I initially had a `TypeError` in that place, but mypy seems to be able to catch the issue, so I was just hoping users would use mypy (plus the docstring already states expect types). but yeah I agree that should be an exception, though I'm a little sad that checks like this are still necessary despite mypy and all type annotations we put in.

Reply:
mypy can help make sure our usage of an API (in other library code or examples, etc.) stays correct, but at the end of the day, outside users can pass in anything they please, so explicit error checking of inputs will always be relevant on public API codepaths.